File: CaddieAI/Services/OverpassClient.swift
//
//  OverpassClient.swift
//  CaddieAI
//
//  HTTP client for OpenStreetMap Overpass API.
//  Used only for fetching detailed golf geometry within a known bounding box.
//  Course search/discovery is handled by NominatimClient.
//

import Foundation

enum OverpassClient {

    // Primary endpoint (no rate limits, fast)
    private static let primaryEndpoint = URL(string: "https://overpass.private.coffee/api/interpreter")!
    // Fallback endpoint
    private static let fallbackEndpoint = URL(string: "https://overpass-api.de/api/interpreter")!

    struct OverpassError: Error, LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    // MARK: - Fetch Golf Features Within Course Area

    static func fetchCourseFeatures(
        boundingBox: CourseBoundingBox
    ) async throws -> OverpassResponse {
        let bbox = boundingBox.buffered(by: 0.002) // ~200m buffer
        let bboxStr = "\(bbox.south),\(bbox.west),\(bbox.north),\(bbox.east)"

        let query = """
        [out:json][timeout:45];
        (
          way["golf"="hole"](\(bboxStr));
          way["golf"="green"](\(bboxStr));
          way["golf"="tee"](\(bboxStr));
          node["golf"="pin"](\(bboxStr));
          way["golf"="bunker"](\(bboxStr));
          way["natural"="water"](\(bboxStr));
          relation["natural"="water"](\(bboxStr));
          way["golf"="fairway"](\(bboxStr));
        );
        out geom;
        """

        return try await executeQueryWithFallback(query)
    }

    // MARK: - Query Execution with Fallback

    private static func executeQueryWithFallback(_ query: String) async throws -> OverpassResponse {
        do {
            return try await executeQuery(query, endpoint: primaryEndpoint)
        } catch {
            // Try fallback endpoint
            return try await executeQuery(query, endpoint: fallbackEndpoint)
        }
    }

    private static func executeQuery(
        _ query: String,
        endpoint: URL,
        maxRetries: Int = 2,
        timeoutSeconds: TimeInterval = 45
    ) async throws -> OverpassResponse {
        var lastError: Error?

        for attempt in 0...maxRetries {
            if attempt > 0 {
                try await Task.sleep(for: .seconds(Double(attempt) * 2))
            }

            do {
                var request = URLRequest(url: endpoint)
                request.httpMethod = "POST"
                request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
                let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
                request.httpBody = "data=\(encoded)".data(using: .utf8)
                request.timeoutInterval = timeoutSeconds

                let (data, response) = try await URLSession.shared.data(for: request)

                guard let http = response as? HTTPURLResponse else {
                    throw OverpassError(message: "Invalid response from Overpass.")
                }

                if http.statusCode == 429 || http.statusCode == 504 {
                    lastError = OverpassError(message: "Overpass rate limited or timed out (HTTP \(http.statusCode)).")
                    continue
                }

                guard http.statusCode == 200 else {
                    throw OverpassError(message: "Overpass API error: HTTP \(http.statusCode)")
                }

                return try JSONDecoder().decode(OverpassResponse.self, from: data)

            } catch let error as OverpassError {
                lastError = error
            } catch is DecodingError {
                throw OverpassError(message: "Could not decode Overpass response.")
            } catch {
                lastError = error
            }
        }

        throw lastError ?? OverpassError(message: "Unknown Overpass error after retries.")
    }
}

File: mobile-flutter/lib/core/courses/overpass_client.dart
// Overpass API client — fetches raw OSM golf-course features for a
// bounding box. Port of ios/CaddieAI/Services/OverpassClient.swift.
//
// Uses two endpoints (primary + fallback) and retries on 429/504 with
// linear backoff. The response is a typed OverpassResponse that the
// OsmParser consumes downstream.

import 'dart:convert';

import 'http_transport.dart';

// ---------------------------------------------------------------------------
// Response model classes
// ---------------------------------------------------------------------------

class OverpassGeomNode {
  final double lat;
  final double lon;
  const OverpassGeomNode({required this.lat, required this.lon});

  factory OverpassGeomNode.fromJson(Map<String, dynamic> j) =>
      OverpassGeomNode(
        lat: (j['lat'] as num).toDouble(),
        lon: (j['lon'] as num).toDouble(),
      );
}

class OverpassMember {
  final String type;
  final int ref;
  final String? role;
  final List<OverpassGeomNode>? geometry;

  const OverpassMember({
    required this.type,
    required this.ref,
    this.role,
    this.geometry,
  });

  factory OverpassMember.fromJson(Map<String, dynamic> j) => OverpassMember(
        type: j['type'] as String,
        ref: (j['ref'] as num).toInt(),
        role: j['role'] as String?,
        geometry: j['geometry'] == null
            ? null
            : (j['geometry'] as List<dynamic>)
                .map((n) =>
                    OverpassGeomNode.fromJson(n as Map<String, dynamic>))
                .toList(growable: false),
      );
}

class OverpassElement {
  final String type;
  final int id;
  final double? lat;
  final double? lon;
  final Map<String, String>? tags;
  final List<OverpassGeomNode>? geometry;
  final List<OverpassMember>? members;

  const OverpassElement({
    required this.type,
    required this.id,
    this.lat,
    this.lon,
    this.tags,
    this.geometry,
    this.members,
  });

  factory OverpassElement.fromJson(Map<String, dynamic> j) => OverpassElement(
        type: j['type'] as String,
        id: (j['id'] as num).toInt(),
        lat: j['lat'] == null ? null : (j['lat'] as num).toDouble(),
        lon: j['lon'] == null ? null : (j['lon'] as num).toDouble(),
        tags: j['tags'] == null
            ? null
            : (j['tags'] as Map<String, dynamic>)
                .map((k, v) => MapEntry(k, v as String)),
        geometry: j['geometry'] == null
            ? null
            : (j['geometry'] as List<dynamic>)
                .map((n) =>
                    OverpassGeomNode.fromJson(n as Map<String, dynamic>))
                .toList(growable: false),
        members: j['members'] == null
            ? null
            : (j['members'] as List<dynamic>)
                .map(
                    (m) => OverpassMember.fromJson(m as Map<String, dynamic>))
                .toList(growable: false),
      );
}

class OverpassResponse {
  final List<OverpassElement> elements;
  const OverpassResponse({required this.elements});

  factory OverpassResponse.fromJson(Map<String, dynamic> j) =>
      OverpassResponse(
        elements: (j['elements'] as List<dynamic>)
            .map((e) => OverpassElement.fromJson(e as Map<String, dynamic>))
            .toList(growable: false),
      );
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

class OverpassClient {
  OverpassClient(this._transport);

  final HttpTransport _transport;

  /// Ordered list of Overpass mirrors to cycle through on failure.
  /// Each mirror has its own rate limiter and server pool, so when
  /// one is rate-limited (typically returning 429 or 504), the others
  /// are likely still healthy. Order is rough reliability rank; the
  /// next request to this client starts from a randomized offset so
  /// load is spread across mirrors over time.
  static const _mirrors = [
    'https://overpass-api.de/api/interpreter',
    'https://overpass.private.coffee/api/interpreter',
    'https://overpass.kumi.systems/api/interpreter',
    'https://overpass.osm.jp/api/interpreter',
    'https://maps.mail.ru/osm/tools/overpass/api/interpreter',
  ];

  static const _bboxBuffer = 0.002;
  static const _maxAttemptsPerMirror = 2;
  static const _timeout = Duration(seconds: 45);

  /// Randomized starting offset so consecutive client instances don't
  /// all hammer the same mirror first.
  static int _mirrorOffset =
      DateTime.now().millisecondsSinceEpoch % _mirrors.length;

  /// Fetches all golf-related OSM features within [south, west, north, east],
  /// with a small buffer added to the bounding box.
  Future<OverpassResponse> fetchCourseFeatures(
    double south,
    double west,
    double north,
    double east,
  ) async {
    final s = south - _bboxBuffer;
    final w = west - _bboxBuffer;
    final n = north + _bboxBuffer;
    final e = east + _bboxBuffer;

    final query = '[out:json][timeout:45];\n'
        '(\n'
        '  way["golf"="hole"]($s,$w,$n,$e);\n'
        '  way["golf"="green"]($s,$w,$n,$e);\n'
        '  way["golf"="tee"]($s,$w,$n,$e);\n'
        '  node["golf"="pin"]($s,$w,$n,$e);\n'
        '  way["golf"="bunker"]($s,$w,$n,$e);\n'
        '  way["natural"="water"]($s,$w,$n,$e);\n'
        '  relation["natural"="water"]($s,$w,$n,$e);\n'
        '  way["golf"="fairway"]($s,$w,$n,$e);\n'
        '  way["leisure"="golf_course"]($s,$w,$n,$e);\n'
        '  relation["leisure"="golf_course"]($s,$w,$n,$e);\n'
        ');\n'
        'out geom;';

    final body = 'data=${Uri.encodeComponent(query)}';

    // Try each mirror in order, starting from a rotating offset.
    // Fast-fail on 429/504 and move to the next mirror immediately.
    Object? lastError;
    final startOffset = _mirrorOffset;
    _mirrorOffset = (_mirrorOffset + 1) % _mirrors.length;

    for (var i = 0; i < _mirrors.length; i++) {
      final endpoint = _mirrors[(startOffset + i) % _mirrors.length];
      try {
        // ignore: avoid_print
        print('OVERPASS: trying $endpoint');
        return await _fetchWithRetry(endpoint, body);
      } catch (e) {
        // ignore: avoid_print
        print('OVERPASS: $endpoint failed: $e');
        lastError = e;
        // Fall through to next mirror.
      }
    }
    throw lastError ??
        Exception('All ${_mirrors.length} Overpass mirrors failed');
  }

  Future<OverpassResponse> _fetchWithRetry(
    String endpoint,
    String body,
  ) async {
    Object? lastError;
    for (var attempt = 1; attempt <= _maxAttemptsPerMirror; attempt++) {
      try {
        final response = await _transport.send(HttpRequestLike(
          method: 'POST',
          url: Uri.parse(endpoint),
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
          body: body,
          timeout: _timeout,
        ));

        // 429/504 → fast-fail this mirror so the caller can try another.
        // Only retry on 504 once within the same mirror (might be a
        // transient queue spike).
        if (response.statusCode == 429) {
          throw Exception('Overpass returned 429 (rate-limited)');
        }
        if (response.statusCode == 504) {
          if (attempt < _maxAttemptsPerMirror) {
            lastError = Exception(
                'Overpass returned 504 on attempt $attempt');
            await Future<void>.delayed(const Duration(seconds: 1));
            continue;
          }
          throw Exception('Overpass returned 504');
        }

        if (!response.isSuccess) {
          throw Exception(
              'Overpass error ${response.statusCode}: ${response.body}');
        }

        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return OverpassResponse.fromJson(json);
      } catch (e) {
        lastError = e;
        if (attempt < _maxAttemptsPerMirror) {
          await Future<void>.delayed(const Duration(seconds: 1));
          continue;
        }
      }
    }
    throw lastError ?? Exception('Overpass request failed');
  }
}
